libraries/platforms/dromajo-vcs/bsg_manycore_simulator.cpp

```diff
@@ -0,0 +1,624 @@
+ * get_argc
+ * Given a string, determine the number of space-separated arguments
+ * @params[in] args - Pointer to a character array that holds the arguments
+ * @returns the number of arguments
+ */
+static int get_argc(char * args){
+        char *cur = args, prev=' ';
+        int count = 1;
+        while(*cur != '\0'){
+                if((prev == ' ') && (prev != *cur)){
+                        count ++;
+                }
+                prev = *cur;
+                ++cur;
+        }
+        return count;
+}
+
+/*
+ * get_argv
+ * Given a string, retrieves the space-separated arguments
+ * @params[in] args - Pointer to a character array that holds the arguments
+ * @params[in] argc - Number of arguments
+ * @params[in] argv - Pointer to an array of strings that will hold the different
+ * arguments
+ */
+static void get_argv(char * args, int argc, char **argv){
+        int count = 0;
+        char *cur = args, prev=' ';
+
+        // First parse the path name. This is not in the argument string because
+        // VCS doesn't provide it to us. Instead, we "hack" around it by reading
+        // the path from 'proc/self/exe'. The maximum path-name length is 1024,
+        // with an extra null character for safety
+        static char path[1025] = {'\0'};
+
+        readlink("/proc/self/exe", path, sizeof(path) - 1);
+        argv[0] = path;
+        count ++;
+
+        // Then we parse the remaining arguments. Arguments are separated by N
+        // >= 1 spaces. We only register an argument when the previous character
+        // was a space, and the current character is not (so that multiple
+        // spaces don't count as multiple arguments). We replace spaces with
+        // null characters (\0) so that each argument appears to be an
+        // individual string and can be used later, by argparse (and other
+        // libraries)
+        while(*cur != '\0'){
+                if((prev == ' ') && (prev != *cur)){
+                        argv[count] = cur;
+                        count++;
+                }
+                prev = *cur;
+                if(*cur == ' ')
+                        *cur = '\0';
+                cur++;
+        }
+}
+
+/*
+ * vcs_main
+ * The main function for the x86 host. It initializes all devices and controls
+ * the time in simulation. It handles all DPI calls and passes packets to and from
+ * Dromajo during execution
+ * @params[in] argc - Number of arguments
+ * @params[in] argv - Pointer to an array of string arguments
+ * @returns
+ */
+int vcs_main(int argc, char **argv) {
+  // Push command-line arguments into global variables
+  _argc = argc;
+  _argv = argv;
+
+  // Initialize Host
+  SimulationWrapper *host = new SimulationWrapper();
+  if (!host) {
+    bsg_pr_err("Could not initialize host!\n");
+    return HB_MC_FAIL;
+  }
+
+  int err;
+  do {
+    err = host->eval();
+    // Codes greater than 0 can be used to terminate a program
+    if (err > 0) break;
+  } while (err >= 0);
```

Comment:
This is the main loop of the program? How does the simulation terminate on successful completion?

Reply:
I do that with the error codes. Every execution of eval() returns a code. HB_MC_SUCCESS if everything was fine or an error code if there was some error. Once the program is complete and needs to terminate successfully, an error code > 0 is returned. I'll add a comment to explain this.

Reply:
Add a TODO as well. We should handle this in some graceful way if we can. It's not for this PR though